### PR TITLE
Feature/model/upload

### PR DESCRIPTION
**Pull request by iroper-tessella**
Upload for models based on the code in the `model-uploader` repo.

I have attempted to produce a helpful error message when a 500 response is returned from the `validate_model_definition` function as it seems as though if the model definition doesn't resemble what it should do, the API does not know how to deal with it.

Tests have been added to all the new functions and the `upload` command added to the `dafni.py` file.

### CODE REVIEW

File: dafni_cli/api/dafni_api.py [hunk 1]
```diff
@@ -51,3 +51,27 @@ def dafni_post_request(
...
+    url: str, jwt: str, data: BinaryIO, content_type: str
+) -> Response:
+    """Performs a PUT request from the DAFNI API.
+    If a status other than 200 is returned, an exception will be raised.
+
+    Args:
+        url (str): The url endpoint that is being queried
+        jwt (str): JWT
+        data (BinaryIO): The data to be PUT-ed
+        content_type (str): Content type for the headers
```
Comment (jamie-short) [inline dafni_cli/api/dafni_api.py:66]: add an example

File: dafni_cli/api/dafni_api.py [hunk 2]
```diff
@@ -51,3 +51,27 @@ def dafni_post_request(
     )
     response.raise_for_status()
     return response.json()
+
+
+def dafni_put_request(
+    url: str, jwt: str, data: BinaryIO, content_type: str
```
Comment (jamie-short) [inline dafni_cli/api/dafni_api.py:57]: might be worth setting a default value?

File: dafni_cli/api/models_api.py [hunk 3]
```diff
@@ -1,8 +1,19 @@
...
-from typing import Union, List
+from typing import Union, List, Tuple, Optional
+from pathlib import Path
+import click
 
-from dafni_cli.consts import MODELS_API_URL, DISCOVERY_API_URL
-from dafni_cli.api.dafni_api import dafni_get_request
+from dafni_cli.consts import (
+    MODELS_API_URL,
+    DISCOVERY_API_URL,
```
Comment (jamie-short) [inline dafni_cli/api/models_api.py:8]: not used

File: dafni_cli/api/models_api.py [hunk 4]
```diff
@@ -39,8 +50,90 @@ def get_model_metadata_dict(jwt: str, model_version_id: str) -> dict:
...
+    """Function to upload definition or image files to DAFNI
+
+    Args:
+        jwt (str): JWT
+        url (str): URL to upload the file to
+        file_path (Path): Path to the file
+    """
+    content_type = MINIO_UPLOAD_CT
+    with open(file_path, "rb") as file_data:
+        dafni_put_request(url, jwt, file_data, content_type)
```
Comment (jamie-short) [inline dafni_cli/api/models_api.py:115]: do we need a `.read()`?

File: dafni_cli/api/models_api.py [hunk 5]
```diff
@@ -39,8 +50,90 @@ def get_model_metadata_dict(jwt: str, model_version_id: str) -> dict:
...
+    """
+    if model_id:
+        url = MODELS_API_URL + "/models/" + model_id + "/upload/" + upload_id + "/ingest/"
+    else:
+        url = MODELS_API_URL + "/models/upload/" + upload_id + "/ingest/"
+    data = {"version_message": version_message}
+    response = dafni_post_request(url, jwt, data)
+
+
+if __name__ == '__main__':
```
Comment (jamie-short) [inline dafni_cli/api/models_api.py:137]: remove

File: dafni_cli/api/models_api.py [hunk 6]
```diff
@@ -39,8 +50,90 @@ def get_model_metadata_dict(jwt: str, model_version_id: str) -> dict:
...
+            upload_id (str): Upload ID
+            version_message (str): Message to be attached to this version
+            model_id (str): ID of existing parent model if it exists
+    """
+    if model_id:
+        url = MODELS_API_URL + "/models/" + model_id + "/upload/" + upload_id + "/ingest/"
+    else:
+        url = MODELS_API_URL + "/models/upload/" + upload_id + "/ingest/"
+    data = {"version_message": version_message}
+    response = dafni_post_request(url, jwt, data)
```
Comment (jamie-short) [inline dafni_cli/api/models_api.py:134]: do we need to return anything?

File: dafni_cli/utils.py [hunk 7]
```diff
@@ -70,4 +70,23 @@ def process_date_filter(date_str: str) -> str:
...
+):
+    """Function to display the arguments and options chosen by the user
+    and ask for confirmation
+
+    Args:
+        argument_names (List[str]): List of the names of the arguments/options for the command
+        arguments (List[str]): List of values chosen for each corresponding argument/option
+        additional_messages (Optional[List[str]]): Other messages to be added after options are listed
+    """
+    for i in range(len(argument_names)):
```
Comment (jamie-short) [inline dafni_cli/utils.py:87]: I always prefer enumerate in these case:
`for i, value in enumerate(argument_names):`
`    click.echo(f"{value}:{arguments[i]}")`

File: dafni_cli/utils.py [hunk 8]
```diff
@@ -70,4 +70,23 @@ def process_date_filter(date_str: str) -> str:
...
+        click.echo(argument_names[i] + ": " + arguments[i])
+    if additional_messages:
+        for message in additional_messages:
+            click.echo(message)
+    click.confirm('Confirm model upload?', abort=True)
```
Comment (jamie-short) [inline dafni_cli/utils.py:92]: good thinking

File: dafni_cli/upload.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@upload.command()
+@click.argument("definition", nargs=1, required=True, type=click.Path(exists=True))
+@click.argument("image", nargs=1, required=True, type=click.Path(exists=True))
+@click.option("--version-message", nargs=1, required=True, type=str)
+@click.option("--parent-model", type=str, default=None)
+@click.pass_context
+def model(
+        ctx: Context, definition: click.Path, image: click.Path, version_message: str, parent_model: str
```
Comment (jamie-short) [inline dafni_cli/upload.py:38]: I would recommend knocking up a simple utils function to validate the paths are actually valid before you do anything with them. something like `os.path.exists(path)`, are we ok with relative paths etc?

**Review by jamie-short — COMMENTED** [review]
I will review tests once the current comments have been addressed/pushed back on

File: dafni_cli/api/dafni_api.py (hunk 1 above)
Comment (iroper-tessella) [reply dafni_cli/api/dafni_api.py:66]: Added

File: dafni_cli/api/dafni_api.py (hunk 2 above)
Comment (iroper-tessella) [reply dafni_cli/api/dafni_api.py:57]: I'm not sure what a good default value would be. It's used twice in the upload, with two different values.
If when uploading datasets etc, there is a `content_type` which is used a lot, then I can add one then.

File: dafni_cli/api/models_api.py (hunk 3 above)
Comment (iroper-tessella) [reply dafni_cli/api/models_api.py:8]: Removed

File: dafni_cli/api/models_api.py (hunk 5 above)
Comment (iroper-tessella) [reply dafni_cli/api/models_api.py:137]: Removed

File: dafni_cli/api/models_api.py (hunk 6 above)
Comment (iroper-tessella) [reply dafni_cli/api/models_api.py:134]: Added

File: dafni_cli/api/dafni_api.py (hunk 2 above)
Comment (jamie-short) [reply dafni_cli/api/dafni_api.py:57]: ah yes, I did see that once I got quite far down. In that case leave as is

File: dafni_cli/upload.py (hunk 9 above)
Comment (iroper-tessella) [reply dafni_cli/upload.py:38]: So Click already has a pretty useful error for this:
<pre>
Usage: runner.py upload model [OPTIONS] DEFINITION IMAGE
Try 'runner.py upload model --help' for help.

Error: Invalid value for 'DEFINITION': Path 'model-file-examples\no_definition.yaml' does not exist.`
</pre>

This comes from the `exists=True` argument. Do you think we need to do a check as well?

File: dafni_cli/utils.py (hunk 7 above)
Comment (iroper-tessella) [reply dafni_cli/utils.py:87]: Changed, and made `confirmation_message` an argument for more flexibility.

File: dafni_cli/upload.py (hunk 9 above)
Comment (jamie-short) [reply dafni_cli/upload.py:38]: ah ok, hadn't come across that. In that case, nope that should suffice

File: test/api/test_dafni_api.py [hunk 10]
```diff
@@ -126,3 +127,55 @@ def test_exception_raised_for_failed_call(
         # ASSERT
         with pytest.raises(HTTPError, match="404 client model"):
             dafni_api.dafni_post_request(url, jwt, data)
+
+
+@patch("dafni_cli.api.dafni_api.requests")
+@patch(
```
Comment (jamie-short) [inline test/api/test_dafni_api.py:133]: is this necessary? As dafni_put_request doesnt use open, so the data object you use in the tests could be anything

File: test/api/test_models_api.py [hunk 11]
```diff
@@ -62,3 +64,166 @@ def test_dafni_get_request_called_correctly(self, mock_get):
...
+    """Test class to test the get_model_upload_urls functionality"""
+
+    def test_post_request_called_with_correct_arguments(self, mock_post):
+        # SETUP
+        jwt = "JWT"
+        url = MODELS_API_URL + "/models/upload/"
+        data = {"image": True, "definition": True}
+
+        # CALL
+        _, _ = models_api.get_model_upload_urls(jwt)
```
Comment (jamie-short) [inline test/api/test_models_api.py:138]: this isnt necessary, you can just do `models_api.get_model_upload_urls(jwt)`

File: test/test_upload.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            # ASSERT
+            assert result.exit_code == 1
+            assert mock_click.echo.call_args_list == [
+                call("Validating model definition"),
+                call("Error validating the model definition. "
+                     "See https://docs.secure.dafni.rl.ac.uk/docs/how-to/models/how-to-write-a-model-definition-file/"
+                     " for guidance")
+            ]
+
+        def test_method_aborted_and_standard_error_printed_if_non200_or_500_response_from_validation_method(
```
Comment (jamie-short) [inline test/test_upload.py:107]: add `_` between non and 200

**Review by jamie-short — COMMENTED** [review]
test coverage looks good

File: test/api/test_dafni_api.py (hunk 10 above)
Comment (iroper-tessella) [reply test/api/test_dafni_api.py:133]: Good point. I've removed it.

File: test/api/test_models_api.py (hunk 11 above)
Comment (iroper-tessella) [reply test/api/test_models_api.py:138]: Changed

File: test/test_upload.py (hunk 12 above)
Comment (iroper-tessella) [reply test/test_upload.py:107]: Added

File: test/api/test_dafni_api.py [hunk 13]
```diff
@@ -1,6 +1,8 @@
 import pytest
-from mock import patch
+from mock import patch, mock_open
```
Comment (jamie-short) [inline test/api/test_dafni_api.py:2]: no longer used

**Pull request merged**